In [1]:
import pandas as pd
import numpy as np

bike_weather = pd.read_csv('bike_weather.csv')
bike_weather

Date_out  Time_out  Count        date  time  cum_precipitation  \
0   2019-10-03         1     64  2019-10-03     1           2.361667   
1   2019-10-03         2     73  2019-10-03     2           3.353333   
2   2019-10-03         3     78  2019-10-03     3           3.930000   
3   2019-10-03         4     57  2019-10-03     4           4.423333   
4   2019-10-03         5     43  2019-10-03     5           4.500000   
..         ...       ...    ...         ...   ...                ...   
64  2019-10-05        19    648  2019-10-05    19           0.600000   
65  2019-10-05        20    619  2019-10-05    20           0.600000   
66  2019-10-05        21    664  2019-10-05    21           0.600000   
67  2019-10-05        22    585  2019-10-05    22           0.600000   
68  2019-10-05        23    418  2019-10-05    23           0.600000   

     humidity  insolation     pressure  sea_lvl_pressure  sunshine       temp  \
0   93.846667        0.00   993.010000       1002.910000         0  20.016667   
1   93.453333        0.00   992.668333       1002.568333         0  19.908333   
2   91.686667        0.00   992.253333       1002.153333         0  19.923333   
3   93.061667        0.00   992.316667       1002.216667         0  19.928333   
4   95.028333        0.00   992.835000       1002.735000         0  19.871667   
..        ...         ...          ...               ...       ...        ...   
64  60.570000        8.24  1008.340000       1018.440000      8400  17.171667   
65  61.381667        8.24  1009.146667       1019.273333      8400  16.610000   
66  62.598333        8.24  1009.773333       1019.973333      8400  15.988333   
67  63.560000        8.24  1010.376667       1020.576667      8400  15.436667   
68  63.426667        8.24  1010.846667       1021.085000      8400  14.986667   

        wind  wind_direction Rain_YN  N  Y  
0   3.290000      178.788333       Y  0  1  
1   3.056667      333.400000       Y  0  1  
2   2.125000      330.110000       Y  0  1  
3   1.931667      251.535000       Y  0  1  
4   2.886667      236.116667       Y  0  1  
..       ...             ...     ... .. ..  
64  4.085000      146.493333       Y  0  1  
65  3.631667      125.785000       Y  0  1  
66  3.686667      225.700000       Y  0  1  
67  3.680000      225.253333       Y  0  1  
68  3.401667      179.688333       Y  0  1  

[69 rows x 17 columns]

In [5]:
bike_weather['over_500'] = 1
bike_weather.loc[bike_weather.Count < 500, 'over_500'] = 0
bike_weather.to_csv('bike_weather.csv', encoding = 'utf-8-sig', index = False)

In [11]:
bike_weather

Date_out  Time_out  Count        date  time  cum_precipitation  \
0   2019-10-03         1     64  2019-10-03     1           2.361667   
1   2019-10-03         2     73  2019-10-03     2           3.353333   
2   2019-10-03         3     78  2019-10-03     3           3.930000   
3   2019-10-03         4     57  2019-10-03     4           4.423333   
4   2019-10-03         5     43  2019-10-03     5           4.500000   
..         ...       ...    ...         ...   ...                ...   
64  2019-10-05        19    648  2019-10-05    19           0.600000   
65  2019-10-05        20    619  2019-10-05    20           0.600000   
66  2019-10-05        21    664  2019-10-05    21           0.600000   
67  2019-10-05        22    585  2019-10-05    22           0.600000   
68  2019-10-05        23    418  2019-10-05    23           0.600000   

     humidity  insolation     pressure  sea_lvl_pressure  sunshine       temp  \
0   93.846667        0.00   993.010000       1002.910000         0  20.016667   
1   93.453333        0.00   992.668333       1002.568333         0  19.908333   
2   91.686667        0.00   992.253333       1002.153333         0  19.923333   
3   93.061667        0.00   992.316667       1002.216667         0  19.928333   
4   95.028333        0.00   992.835000       1002.735000         0  19.871667   
..        ...         ...          ...               ...       ...        ...   
64  60.570000        8.24  1008.340000       1018.440000      8400  17.171667   
65  61.381667        8.24  1009.146667       1019.273333      8400  16.610000   
66  62.598333        8.24  1009.773333       1019.973333      8400  15.988333   
67  63.560000        8.24  1010.376667       1020.576667      8400  15.436667   
68  63.426667        8.24  1010.846667       1021.085000      8400  14.986667   

        wind  wind_direction Rain_YN  N  Y  over_500  
0   3.290000      178.788333       Y  0  1         0  
1   3.056667      333.400000       Y  0  1         0  
2   2.125000      330.110000       Y  0  1         0  
3   1.931667      251.535000       Y  0  1         0  
4   2.886667      236.116667       Y  0  1         0  
..       ...             ...     ... .. ..       ...  
64  4.085000      146.493333       Y  0  1         1  
65  3.631667      125.785000       Y  0  1         1  
66  3.686667      225.700000       Y  0  1         1  
67  3.680000      225.253333       Y  0  1         1  
68  3.401667      179.688333       Y  0  1         0  

[69 rows x 18 columns]

In [12]:
from sklearn.model_selection import train_test_split

X = bike_weather[['cum_precipitation', 'humidity', 'temp', 'wind']]
y = bike_weather.over_500
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 123)

In [13]:
import statsmodels.api as sm

X1 = sm.add_constant(X_train)
logit_model = sm.Logit(y_train, X1)
result = logit_model.fit()
print(result.summary())

C:\Users\kissd\anaconda3\lib\site-packages\statsmodels\tsa\tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


Optimization terminated successfully.
         Current function value: 0.392092
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:               over_500   No. Observations:                   48
Model:                          Logit   Df Residuals:                       43
Method:                           MLE   Df Model:                            4
Date:                Fri, 13 Jan 2023   Pseudo R-squ.:                  0.4227
Time:                        22:07:13   Log-Likelihood:                -18.820
converged:                       True   LL-Null:                       -32.601
Covariance Type:            nonrobust   LLR p-value:                 1.530e-05
                        coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                11.4773      8.488      1.352      0.176      -5.158      28.113
cum_prec

Pseudo R-squ : 회귀분석이 아니므로 Pseudo라는 표현 사용, 회귀처럼 계산하면 0.4227 정도의 R제곱값이 나온다는 의미  
LLR p-value : 1.530e-05이므로 아주 작은 값 : 통계적 유의성에 문제없음  
독립변수들의 p-value : humidity를 제외한 독립변수들은 통계성 유의성이 부족함  

유의성이 확인되지 않은 독립변수  
일반적인 통계에서는 p-value의 유의성이 확인되지 않은 독립변수를 제외  
> 머신러닝에서는 시각이 다름  

In [14]:
from sklearn.linear_model import LogisticRegression

log_reg = LogisticRegression()
log_reg.fit(X_train, y_train)
print('Train set 정확도 : %.2f'%log_reg.score(X_train, y_train))
print('Test set 정확도 : %.2f'%log_reg.score(X_test, y_test))

Train set 정확도 : 0.88
Test set 정확도 : 0.76


In [16]:
from sklearn.metrics import classification_report

y_pred = log_reg.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.70      0.78      0.74         9
           1       0.82      0.75      0.78        12

    accuracy                           0.76        21
   macro avg       0.76      0.76      0.76        21
weighted avg       0.77      0.76      0.76        21

